lectures/04_lecture/data_pipelines.ipynb

```diff
@@ -0,0 +1,1573 @@
+       "MLPClassifier(hidden_layer_sizes=6)                 0.813091\n",
+       "MLPClassifier(hidden_layer_sizes=10)                0.865318\n",
+       "MLPClassifier(hidden_layer_sizes=20)                0.872352\n",
+       "MLPClassifier(hidden_layer_sizes=40)                0.876567"
+      ]
+     },
+     "execution_count": 21,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from sklearn.metrics import f1_score\n",
+    "\n",
+    "results = {}\n",
+    "for model in models_to_run:\n",
+    "\n",
+    "    # create instance of the class, feed the model to be tested\n",
+    "    clf = make_pipeline(categorical_cols, numerical_cols, model)\n",
+    "    clf.fit(X_train_wine, y_train_wine)\n",
+    "\n",
+    "    # store results in a dictionary \n",
+    "    results[str(model)] = f1_score(y_test_wine, clf.predict(X_test_wine), average='weighted')\n",
+    "pd.DataFrame.from_dict(results, orient='index', columns=['score'])\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Unfortunately the parameter did not change our model for the better. This can ofcourse always happen in Machine Learning, which is sometimes more of an art than a science. Several other methods to combat class imbalance could be checked for (oversampling, smote...), but an increase in performance is never guaruanteed. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## IN PRACTICE: MNIST DATASET"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will turn to a second problem: the classification of handwritten digits. This is a computer vision task, and we know that Logistic Regression and Random Forest are ill-equipped for this task. Instead, we will add a convolutional neural network to our pipeline."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from tensorflow.keras.datasets.mnist import load_data\n",
+    "from tensorflow.keras import Sequential\n",
+    "from tensorflow.keras.layers import Conv2D\n",
+    "from tensorflow.keras.layers import MaxPool2D\n",
+    "from tensorflow.keras.layers import Dense\n",
+    "from tensorflow.keras.layers import Flatten\n",
+    "from tensorflow.keras.layers import Dropout\n",
+    "from tensorflow.keras.utils import plot_model\n",
+    "from tensorflow.keras.datasets.mnist import load_data\n",
+    "from tensorflow.keras.wrappers.scikit_learn import KerasClassifier"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.datasets import load_digits"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "(X_digits_train, y_digits_train), (X_digits_test, y_digits_test) = load_data()"
```

Comment:
Is this necessary? I would be affraid that our students will be be confused by this! 😄  

I would write this:

```suggestion
    "X_digits_train, y_digits_train, X_digits_test, y_digits_test = load_data()"
```

But I don't know.